fusequery/query/src/interpreters/plan_scheduler.rs

```diff
@@ -2,380 +2,860 @@
+        }
+    }
+
+    fn visit_broadcast(&mut self, plan: &BroadcastPlan, tasks: &mut Tasks) -> Result<()> {
+        self.visit_plan_node(plan.input.as_ref(), tasks)?;
+
+        // Entering new stage
+        self.stage_id = uuid::Uuid::new_v4().to_string();
+
+        match self.running_mode {
+            RunningMode::Cluster => self.visit_cluster_broadcast(tasks),
+            RunningMode::Standalone => self.visit_local_broadcast(tasks),
+        };
+
+        Ok(())
+    }
+
+    fn broadcast_action(&self, input: &PlanNode) -> BroadcastAction {
+        BroadcastAction {
+            stage_id: self.stage_id.clone(),
+            query_id: self.query_context.get_id(),
+            plan: input.clone(),
+            sinks: self.cluster_nodes.clone(),
+        }
+    }
+
+    fn broadcast_remote(&self, node_name: &str, action: &BroadcastAction) -> RemotePlan {
+        RemotePlan {
+            schema: action.plan.schema(),
+            query_id: action.query_id.clone(),
+            stage_id: action.stage_id.clone(),
+            stream_id: node_name.to_string(),
+            fetch_nodes: self.cluster_nodes.clone(),
+        }
+    }
+
+    fn visit_local_broadcast(&mut self, tasks: &mut Tasks) {
+        self.running_mode = RunningMode::Cluster;
+        let node_name = &self.cluster_nodes[self.local_pos];
+        let action = self.broadcast_action(&self.nodes_plan[self.local_pos]);
+        tasks.add_task(node_name, FlightAction::BroadcastAction(action.clone()));
+
+        for index in 0..self.nodes_plan.len() {
+            let node_name = &self.cluster_nodes[index];
+            self.nodes_plan[index] = PlanNode::Remote(RemotePlan {
+                schema: action.plan.schema(),
+                query_id: action.query_id.clone(),
+                stage_id: action.stage_id.clone(),
+                stream_id: node_name.to_string(),
+                fetch_nodes: vec![self.cluster_nodes[self.local_pos].clone()],
+            });
+        }
+    }
+
+    fn visit_cluster_broadcast(&mut self, tasks: &mut Tasks) {
+        self.running_mode = RunningMode::Cluster;
+        for index in 0..self.nodes_plan.len() {
+            let node_name = &self.cluster_nodes[index];
+            let action = self.broadcast_action(&self.nodes_plan[index]);
+            let remote_plan_node = self.broadcast_remote(node_name, &action);
+
+            tasks.add_task(node_name, FlightAction::BroadcastAction(action));
+            self.nodes_plan[index] = PlanNode::Remote(remote_plan_node);
+        }
+    }
+
+    fn visit_projection(&mut self, plan: &ProjectionPlan, tasks: &mut Tasks) -> Result<()> {
+        self.visit_plan_node(plan.input.as_ref(), tasks)?;
+        match self.running_mode {
+            RunningMode::Cluster => self.visit_cluster_projection(plan),
+            RunningMode::Standalone => self.visit_local_projection(plan),
+        };
+        Ok(())
+    }
+
+    fn visit_local_projection(&mut self, plan: &ProjectionPlan) {
```

Comment:
Can we abstract a method `get_pos_by_mod(&self) -> Vec<usize>` ?

Thus we don't need to keep cluster and local functions.

Reply:
Let's change it in another pr.